src/Table.hxx

```diff
@@ -108,6 +108,62 @@ public:
 
     std::vector<std::pair<std::string, std::string>> flatten_properties() const;
 
+    //====================================================================
+
+    inline std::map<size_t, Data_Type> determine_datatypes_for_writing(
+            const Format &format) const {
+        std::map<size_t, Data_Type> alt_datatype_lookup;
+        bool adjust_uint64 = ((format.enum_format == Format::Enums::VOTABLE) ||
+                              (format.enum_format == Format::Enums::IPAC_TABLE) ||
+                              (format.enum_format == Format::Enums::FITS));
+        bool adjust_all_unsigned = (format.enum_format == Format::Enums::VOTABLE);
+
+        // Skip the null bitfield flag
+        for (size_t col = 1; col <= columns.size(); ++col) {
+            if ((columns[col].type == Data_Type::UINT64_LE) && adjust_uint64) {
+                if (contains_large_uint64_val(col)) {
+                    alt_datatype_lookup[col] = Data_Type::CHAR;
+                } else {
+                    alt_datatype_lookup[col] = Data_Type::INT64_LE;
+                }
+            } else if (adjust_all_unsigned) {
+                if (columns[col].type == Data_Type::UINT32_LE) {
+                    alt_datatype_lookup[col] = Data_Type::INT64_LE;
+                } else if (columns[col].type == Data_Type::UINT16_LE) {
+                    alt_datatype_lookup[col] = Data_Type::INT32_LE;
+                }
+            }
+        }
+        return alt_datatype_lookup;
+    }
+
+    inline bool contains_large_uint64_val(size_t col) const {
+        for (size_t table_row_offset(0), row(0); table_row_offset < data.size();
+             table_row_offset += row_size(), ++row) {
+            auto curr_ptr = data.data() + table_row_offset + offsets[col];
+            auto array_size = columns[col].array_size;
+            for (size_t j = 0; j < array_size; ++j) {
+                uint64_t val = *reinterpret_cast<const uint64_t *>(curr_ptr);
+                if (val > std::numeric_limits<int64_t>::max()) {
+                    return true;
+                }
+                curr_ptr += sizeof(uint64_t);
+            }
+        }
+        return false;
+    }
+
+    inline Data_Type get_datatype_for_writing(
+            const std::map<size_t, Data_Type> &datatypes_for_writing, size_t col) const {
+        auto iter = datatypes_for_writing.find(col);
+        if (iter == datatypes_for_writing.end()) {
+            return columns[col].type;
+        }
+        return iter->second;
+    }
```

Comment:
Instead of a map from column index to datatypes, I think it would probably be simpler and more performant to use a vector of output datatypes.

Reply:
Okay.  For performance, I'm more worried about iterating through great big tables.